include/aspect/material_model/cpp.h

```diff
@@ -0,0 +1,215 @@
+/*
+  Copyright (C) 2018 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef _aspect_material_model_cpp_h
+#define _aspect_material_model_cpp_h
+
+#include <aspect/material_model/interface.h>
+#include <aspect/simulator_access.h>
+#include <aspect/utilities.h>
+#include <regex>
+
+// This plugin requires that we can load shared libraries.
+#if ASPECT_USE_SHARED_LIBS==1
+#include <dlfcn.h>
+#endif
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    using namespace dealii;
+
+    /**
+     * A structure for organizing user code snippets.
+     */
+    struct UserCode
```

Comment:
Can you put this into a namespace specific to the CPP material model? Like `aspect::MaterialModel::internal::CPP::UserCode`? This is lengthy, but hides the rather generic name from all other uses in the namespace.

Reply:
Alternatively, and maybe better, why not put it into the material model itself? Or are these global parameters in the `.prm` file?